rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+                    # FIXME: seems we can just remove entities here....
+            elif isinstance(event, ActionExecuted):
+                sub_states_from_event = []
+                sub_state = event.as_sub_state()
+                for key in [ACTION_NAME, ACTION_TEXT]:
+                    if key in sub_state:
+                        sub_states_from_event.append({key: sub_state[key]})
+            else:
+                sub_states_from_event = []
+
+            for sub_state in sub_states_from_event:
+                message = Message(data=sub_state)
+                self.add(message)
+
+
+class CoreFeaturizationPreparer(GraphComponent):
+    """Provides training data for core's NLU pipeline as well as lookup table buildup.
+
+    During training as well as during inference, the given data (i.e. story graph or
+    list of messages), will be used to derive a de-duplicated list of all possible
+    "prev_action" and "user" sub-states (with either `TEXT` and possibly `ENTITY`
+    attributes or only `INTENT` attribute) that could be generated from the given data.
+    """
+
+    # NOTE: If we wanted to create the training data for the components of the e2e
+    # featurization pipeline differently from the data that is needed for the lookup
+    # table (currently, the latter is used for training as well), we simply need to
+    # move the functionality of this component to e.g. a `StoryToLookUpDataConverter`,
+    # replace the functionality here with the required training data build up, and
+    # adapt the graph connections so that the `train` and `predict` steps of the e2e
+    # run consecutively and use the respective data.
+
+    @classmethod
+    @abstractmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> GraphComponent:
+        """Creates a new `CoreFeaturizationPreparer`.
+
+        Args:
+            config: unused
+            model_storage: unused
+            resource: unused
+            execution_context: unused
+
+        Returns: An instantiated `CoreFeaturizationPreparer`.
+        """
+        return cls()
+
+    WORKAROUND_TEXT = "hi"
+
+    @staticmethod
+    def prepare_for_training(domain: Domain, story_graph: StoryGraph,) -> TrainingData:
+        """Creates a list of unique (partial) substates from the domain and story graph.
+
+        Note that partial user substate means user substate with intent only, user
+        substate with all attributes expect intent, "prev_action" substate with
+        action text only or "prev_action" substate with action name only.
+
+        Args:
+           domain: the domain
+           story_graph: a story graph
+        Returns:
+           training data for core's NLU pipeline
+        """
+        lookup_table = CoreFeaturizationPrecomputations(handle_collisions=True)
```

Comment:
This was meant for the case that you annotated instance number 1 of some text with entities and forgot to annotate instance number 2 of the same text in the same way. The lookup table would only remember one of the results because it's using the text as key. .... Don't think about it too much. If that happens I guess the better solution would be to use keys including the entities (which we can look up). But it should be irrelevant, there should be no entities needed (I think I just messed up when trying to remove those and thought I had overlooked some kind of precomputation wrt entities).

Reply:
Thanks for the explanation! Why are not entities needed?

Reply:
Cause the tokenize+featurize pipeline should not add anything to entities or behave differently if entities are present - I think. The state featurizer just needs to be able to lookup the result of the tokenization of the TEXT when featurizing the entit_data later (I think/hope).

Reply:
oh I see. Great point!